src/main/java/org/commcare/core/process/CommCareInstanceInitializer.java

```diff
@@ -108,7 +109,7 @@ public InstanceRoot generateRoot(ExternalDataInstance instance) {
      */
     protected InstanceRoot setupExternalDataInstance(ExternalDataInstance instance, String reference,
             String stepType) {
-        String referenceId = getRefId(reference);
+        String referenceId = VirtualInstances.getReferenceId(reference);
         InstanceRoot instanceRoot = getExternalDataInstanceSource(instance, referenceId, stepType);
```

Comment:
I don't think this will ever return null. I think the last line in `getExternalDataInstanceSource` should be changed to `return null` and this function should change to:

```java
if (instanceRoot != null) {
    return instanceRoot;
}
return instance.getSource() == null ? ConcreteInstanceRoot.NULL : instance.getSource();